sstables_loader.cc

```diff
@@ -192,7 +195,18 @@ class tablet_sstable_streamer : public sstable_streamer {
 };
 
 inet_address_vector_replica_set sstable_streamer::get_endpoints(const dht::token& token) const {
-    return get_all_endpoints(token);
+    return get_all_endpoints(token) | std::views::filter([&topo = _erm->get_topology(), scope = _stream_scope] (const auto& ep) {
+        switch (scope) {
+        case stream_scope::all:
+            return true;
+        case stream_scope::dc:
+            return topo.get_datacenter(ep) == topo.get_datacenter();
```

Comment:
Shouldn't scope interact with --primary-replica-only?
It makes sense to stream to one replica in the dc and then repair each dc to rebuild the other replicas.

It's less useful per rack since first, we expect one replica per rack at most in the typical case, and repair/rebuild have a source_dc parameter to limit their scope, not source_rack.

Reply:
Isn't primary-replica-only a scope?

Reply:
> Isn't primary-replica-only a scope?

It's possible but then I think we'll need also primary-replica-in-dc scope since the current option is global.

Reply:
> Isn't primary-replica-only a scope?

These two are different, but the term "scope" can be applied to primary-replica-only too, I suppose.
E.g. -- if scope==node and primary-replica is set and current node is not "primary", mutation will be skipped

Reply:
@xemul I'm sitting on the fence WRT primary-replica-only being a distinct scope.
I think it's ok to keep it as an orthogonal option, but in this case, it should interact with the scope since currently it refers to a global primary replica.
That might have made sense with SimpleStrategy, but with NetworkTopologyStrategy, I think it should be applied to a primary replica in the local datacenter, if the `dc` scope is used.
With `rack` scope, we can also define as as a primary (first in ordinal order) in rack, if there are multiple replicas in the rack.
With `node`, as you suggested, it can limit streaming in the local node only if it is the primary replica (global? in-dc? maybe decide that as a function of the keyspace replication strategy?)

Reply:
> @xemul I'm sitting on the fence WRT primary-replica-only being a distinct scope.

And not only you :) I don't mind moving to your fence, but so far I don't understand your motivation, apparently I'm missing something ~~obvious~~, what is it?

> That might have made sense with SimpleStrategy, but with NetworkTopologyStrategy, I think it should be applied to a primary replica in the local datacenter, if the dc scope is used.
> With rack scope, we can also define as as a primary (first in ordinal order) in rack, if there are multiple replicas in the rack.
> With node, as you suggested, it can limit streaming in the local node only if it is the primary replica (global? in-dc? maybe decide that as a function of the keyspace replication strategy?)

The scope motivation is to parallelize load-and-stream between "scope domains".

Consider scope==dc. Currently, you can call load-and-stream on _any_ node in the cluster and it will spread the mutations over all DCs. With scope, you can do the same by picking one node from _each_ dc and call load-and-stream on them all. And it will (will it?) work regardless of whether primary-replica-only is set or not.

The same applies to scope=rack, provided with scope set nodes are picked from each rack. And the corner case -- calling load-and-stream on each node from the cluster with scope=node also works.

I accept Raphael's concern about tablets migration that can happen in parallel and mess replicas up, is it the only problem out there?

Reply:
> > Isn't primary-replica-only a scope?
> 
> These two are different, but the term "scope" can be applied to primary-replica-only too, I suppose. E.g. -- if scope==node and primary-replica is set and current node is not "primary", mutation will be skipped

I'm trying to reduce the number of terms the user has to understand. If we can fold primary-replica-only into a new concept scope that replaces it, there are fewer things to understand.

If we want primary-replica-only-and-this-node-only, that reduces the advantage of folding the two concepts together, but do we want primary-replica-only-and-this-node-only?

Reply:
The primary replica option limits the **number** of replicas data is streamed to to one, not the locality of streaming.  Therefore, its semantics can be defined as:
- no scope: send only to global primary replica
- dc: send only to primary replica in this dc
- rack: send only to primary replica in the dc, if it's in this rack
- node: send only to primary replica in the dc, if it's this node

Reply:
@xemul will we rely on user disabling migration during load-and-stream with scope? this is a different behavior than what we have today without this new scope param. I think we should at least enforce it by failing the command if active migration (== there is a pending replica for a given token) is detected.

Reply:
> Holding the erm should be enough, same as regular writes. That said, if the a tablet is pending migrations we should do double writes and stream to both source and destination.

I don't it's enough, since migrations may have started before we decided to run load-and-stream. We don't do the double write, if we filter out the pending replica as we're doing here (that's how I interpreted the changes, am I wrong?).

Reply:
> > Holding the erm should be enough, same as regular writes. That said, if the a tablet is pending migrations we should do double writes and stream to both source and destination.
> 
> I don't it's enough, since migrations may have started before we decided to run load-and-stream. We don't do the double write, if we filter out the pending replica as we're doing here (that's how I interpreted the changes, am I wrong?).

But the tablet_map provided by the erm will include all pending replicas that are currently being migrated to.  Regular writes wouldn't work with parallel migrations otherwise.

Cc @tgrabiec

Reply:
> > > Holding the erm should be enough, same as regular writes. That said, if the a tablet is pending migrations we should do double writes and stream to both source and destination.
> > 
> > 
> > I don't it's enough, since migrations may have started before we decided to run load-and-stream. We don't do the double write, if we filter out the pending replica as we're doing here (that's how I interpreted the changes, am I wrong?).
> 
> But the tablet_map provided by the erm will include all pending replicas that are currently being migrated to. Regular writes wouldn't work with parallel migrations otherwise.
> 
> Cc @tgrabiec

It does provide, but through different calls.

See for example how storage_proxy::create_write_response_handler_helper() combines natural endpoints + pending.

tablet_effective_replication_map's get_natural_endpoints() -> get_replicas_for_write()

If migration stage is write both, we return current replica set. Then get_pending_endpoints() retrieves pending replica.

Please see the other comment of me explaining how I implemented load-and-stream ability to work with active migrations. It wasn't working before that. This patch breaks it.

Reply:
@bhalevy This is the commit b69ee6bc27bec54c085681b239b9d53633f0ecf2

Reply:
I just explained above how we can avoid breaking it.

Reply:
It'd be best if we had a test for it to demonstrate the breakage and verify the fix

Reply:
Sounds good. I think we can extend existing 

```
@pytest.mark.parametrize("primary_replica_only", [False, True])
async def test_tablet_load_and_stream(manager: ManagerClient, primary_replica_only):
```

It stresses concurrent migration in parallel to load-and-stream.